src/commonMain/kotlin/fr/acinq/lightning/channel/Commitments.kt

```diff
@@ -764,43 +792,91 @@ data class Commitments(
-                val commitments1 = copy(active = active.map {
-                    if (it.fundingTxId == txId) {
-                        log.info { "setting localFundingStatus=${status::class.simpleName} for funding txid=$txId" }
-                        it.copy(localFundingStatus = status)
-                    } else {
-                        it
-                    }
-                }).pruneCommitments(log)
-                val commitment = commitments1.active.find { it.fundingTxId == txId }!!
-                Either.Right(Pair(commitments1, commitment))
+                logger.warning { "fundingTxId=$fundingTxId doesn't match any of our funding txs" }
+                Either.Left(this@Commitments)
+            }
+        }
+    }
+
+    fun ChannelContext.updateLocalFundingStatus(fundingTxId: ByteVector32, status: LocalFundingStatus): Either<Commitments, Pair<Commitments, Commitment>> =
+        updateFundingStatus(fundingTxId) { c: Commitment, _: Long ->
+            if (c.fundingTxId == fundingTxId) {
+                logger.debug { "setting localFundingStatus=${status::class.simpleName} for funding fundingTxId=$fundingTxId" }
+                c.copy(localFundingStatus = status)
+            } else c
+        }
+
+    fun ChannelContext.updateRemoteFundingStatus(fundingTxId: ByteVector32): Either<Commitments, Pair<Commitments, Commitment>> =
+        updateFundingStatus(fundingTxId) { c: Commitment, fundingTxIndex: Long ->
+            // all funding older than this one are considered locked
+            if (c.fundingTxId == fundingTxId || c.fundingTxIndex < fundingTxIndex) {
+                logger.debug { "setting remoteFundingStatus=${RemoteFundingStatus.Locked::class.simpleName} for fundingTxId=$fundingTxId" }
+                c.copy(remoteFundingStatus = RemoteFundingStatus.Locked)
+            } else c
+        }
+
+    /**
+     * Commitments are considered inactive when they have been superseded by a newer commitment, but can still potentially
+     * end up on-chain. This is a consequence of using zero-conf. Inactive commitments will be cleaned up by
+     * [pruneCommitments], when the next funding tx confirms.
+     */
+    private fun ChannelContext.deactivateCommitments(): Commitments {
+        val lastLocalLockedIndex: Long = active.firstOrNull { staticParams.useZeroConf || it.localFundingStatus is LocalFundingStatus.ConfirmedFundingTx }?.fundingTxIndex ?: -1
+        val lastRemoteLockedIndex: Long = active.firstOrNull { it.remoteFundingStatus is RemoteFundingStatus.Locked }?.fundingTxIndex ?: -1
+        val lastLockedIndex = min(lastLocalLockedIndex, lastRemoteLockedIndex)
+        return when (val lastLocked = active.find { it.fundingTxIndex == lastLockedIndex }) {
+            is Commitment -> {
+                // all commitments older than this one are inactive
+                val inactive1 = active.filter { it.fundingTxId != lastLocked.fundingTxId && it.fundingTxIndex <= lastLocked.fundingTxIndex }
```

Comment:
I was wrong, we cannot use `<` here otherwise we won't correctly prune RBF attempts when we're *not* using 0-conf, because we only prune _inactive_ commitments. But we really need to make sure we never RBF a transaction using 0-conf, otherwise we can mess up our commitment state!